### PR TITLE
Add the bouncer-networks extension specification

### PR DESCRIPTION
**Pull request by emersion**
This is based on a previous proposal by Darren Whitlen [1], but with
its scope reduced to networks (buffers are left out). Discussion has
happened in an ircv3-ideas issue [2].

[1]: https://gist.github.com/prawnsalad/d864822d9bb8116626a4370bee6e6fc6
[2]: https://github.com/ircv3/ircv3-ideas/issues/39

cc @prawnsalad

### CODE REVIEW

File: extensions/bouncer-networks.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    BOUNCER LISTNETWORKS
+
+The server replies with:
+
+    BOUNCER LISTNETWORKS <netid> <attributes>
+
+And then ends the list (even if empty) with:
+
+    BOUNCER LISTNETWORKS RPL_OK
```
Comment (emersion) [inline extensions/bouncer-networks.md:74]: Not a fan if this `RPL_OK` token. Something like `*` would probably be better.

**Comment by jwheare** [comment]
It's not immediately clear from this issue why the buffers bits were left out of this, care to elaborate @emersion?

@prawnsalad any thoughts? Are you keen to standardise the buffers portion of your original proposal as well, or could that be part of a later or vendor-specific extension?

It's a question of general applicability I suppose. One issue is probably that servers don't really have the same concept of "buffers" as clients, but bouncers often do. See https://wiki.znc.in/Query_buffers

**Comment by emersion** [comment]
My bouncer doesn't have a concept of "buffers", which is why these are left out. I guess the reason why ZNC has them is because they store history in memory and need to know when it's safe to release the memory. Since soju stores history on disk, there's no need for this.

**Comment by jwheare** [comment]
Aside from memory management (which seems a red herring to me) I think this disconnect directly relates to the same questions coming up in chathistory around history for private messages. Servers don't have a concept of them, but I think there's a stronger case for it with bouncers, as they operate as both client and server at the same time and need to model more of what the client does.

IRCCloud has a pm buffers concept, as does Quassel, and not for memory management reasons.

**Comment by emersion** [comment]
> IRCCloud has a pm buffers concept, as does Quassel, and not for memory management reasons.

OK, what are they used for then?

**Comment by jwheare** [comment]
They map to the way clients work, messages are grouped within buffers, and requested by the client as an ordered list within the concept of the buffer, so the bouncer has to be able to send them that way, With a shared buffer for all PMs, you get the issue that spam in one PM pushes out legitimate messages in another.

**Comment by emersion** [comment]
Hm, this still looks like this is still tied to history retrieval? Why is a concept of buffer necessary for this? soju doesn't have a concept of buffer, but still has per-PM history buckets and doesn't run into the spam issue you're describing.

In any case, as said earlier, adding another extension on top of this one for bouncer buffers should be pretty trivial. I don't think we should force bouncers to implement a concept of buffers because of this extension.

**Comment by jwheare** [comment]
> has per-PM history buckets

That's basically all I mean by a buffer. I don't mean a literal memory buffer.

> I don't think we should force bouncers to implement a concept of buffers because of this extension.

That's fair, I just wonder if the resistance to treating PM buckets as a first class concept will continue to lead to issues as servers begin to implement bouncer type features (history, notifications, etc)

I also just think this point needed to made explicit in this issue.

**Comment by slingamn** [comment]
>Servers don't have a concept of them, but I think there's a stronger case for it with bouncers, as they operate as both client and server at the same time and need to model more of what the client does. [...] With a shared buffer for all PMs, you get the issue that spam in one PM pushes out legitimate messages in another.

This is subtle, I think. CHATHISTORY mandates that servers be able to retrieve PMs with a specific correspondent (nick targets), as well as with all correspondents (the `*` target). What CHATHISTORY does not provide is a way to enumerate all the "open queries" (open buffers?), since this concept is time-relative on the server side.

**Comment by jwheare** [comment]
Yeah but it's still an issue because you have to know that a nick pmed you to request the chat from it. The * target is the workaround for that, but it has the issue I described.

**Comment by slingamn** [comment]
That's what I meant by "open queries". I'm interested in potential solutions to this problem.

**Comment by emersion** [comment]
WIP implementations are available in the `bouncer-ext` branches of soju and gamja:

- https://git.sr.ht/~emersion/soju/tree/bouncer-ext
- https://git.sr.ht/~emersion/gamja/tree/bouncer-ext

File: extensions/bouncer-networks.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+    BOUNCER LISTNETWORKS
+
+The server replies with:
+
+    BOUNCER LISTNETWORKS <netid> <attributes>
+
+And then ends the list (even if empty) with:
+
+    BOUNCER LISTNETWORKS *
```
Comment (emersion) [inline extensions/bouncer-networks.md:74]: It may be desirable to use a batch instead of this special end-of-list message.

**Comment by prawnsalad** [comment]
I don't think I'm following why you would want to remove buffer support from the spec? I assume your bouncer has a list of joined channels and/or queries either in memory or on disk? If so, these would be your buffers.

If you don't have a list of joined channels, how does your bouncer handle notifying the client of joined channels?

**Comment by emersion** [comment]
I have a list of joined channels in memory, but I don't have a list of queries. It's not useful for my use-case.

The client is notified of the joined channels via regular JOIN messages. The client can manage channel buffers via regular JOIN/PART commands.

**Comment by DanielOaks** [comment]
I don't see a reason to leave the buffers part from the original spec out.

**Comment by emersion** [comment]
I don't see a reason to implement it.

Alright, I'll just ship my vendored ext. I'll reconsider submitting an IRCv3 proposal once things have settled.

**Comment by emersion** [comment]
For future reference, I'll keep an up-to-date spec for the vendored ext here: https://git.sr.ht/~emersion/soju/tree/bouncer-ext/item/doc/ext/bouncer-networks.md

**Comment by ilmari-lauhakangas** [comment]
> For future reference, I'll keep an up-to-date spec for the vendored ext here: https://git.sr.ht/~emersion/soju/tree/bouncer-ext/item/doc/ext/bouncer-networks.md

New location: https://git.sr.ht/~emersion/soju/tree/master/item/doc/ext/bouncer-networks.md

**Pull request closed without merging**